qiita_ware/ebi.py

```diff
@@ -591,231 +528,57 @@ def generate_submission_xml(self, action):
         actions = ET.SubElement(submission, 'ACTIONS')
 
         study_action = ET.SubElement(actions, 'ACTION')
-        ET.SubElement(study_action, action, {
+        ET.SubElement(study_action, self.action, {
             'schema': 'study',
             'source': basename(self.study_xml_fp)}
         )
 
         sample_action = ET.SubElement(actions, 'ACTION')
-        ET.SubElement(sample_action, action, {
+        ET.SubElement(sample_action, self.action, {
             'schema': 'sample',
             'source': basename(self.sample_xml_fp)}
         )
 
         experiment_action = ET.SubElement(actions, 'ACTION')
-        ET.SubElement(experiment_action, action, {
+        ET.SubElement(experiment_action, self.action, {
             'schema': 'experiment',
             'source': basename(self.experiment_xml_fp)}
         )
 
         run_action = ET.SubElement(actions, 'ACTION')
-        ET.SubElement(run_action, action, {
+        ET.SubElement(run_action, self.action, {
             'schema': 'run', 'source': basename(self.run_xml_fp)}
         )
 
-        if action == 'ADD':
+        if self.action == 'ADD':
             hold_action = ET.SubElement(actions, 'ACTION')
             ET.SubElement(hold_action, 'HOLD', {
-                'HoldUntilDate': str(date.today() + timedelta(365))}
+                'HoldUntilDate': str(date_to_hold + timedelta(365))}
             )
 
-        return submission_set
+        return ET.tostring(submission_set)
 
-    def _write_xml_file(self, xml_gen_fn, attribute_name, fp,
-                        xml_gen_fn_arg=None):
+    def write_xml_file(self, text, attribute_name, fp):
         """Writes an XML file after calling one of the XML generation
         functions
 
         Parameters
         ----------
-        xml_gen_fn : function
-            The function that will be called to generate the XML that will be
-            written
+        text : str
+            The XML text that will be written
         attribute_name : str
             The name of the attribute in which to store the output filepath
         fp : str
             The filepath to which the XML will be written
-        xml_gen_fn_arg : str, optional
-            Defaults to None. If None, no arguments will be passed to
-            xml_gen_fn. Otherwise, this will be passed as the only argument to
-            xml_gen_fn
-
-        Notes
-        -----
-        xml_gen_fn_arg is needed for generating the submission XML
         """
-        if xml_gen_fn_arg is None:
-            xml_element = xml_gen_fn()
-        else:
-            xml_element = xml_gen_fn(xml_gen_fn_arg)
-
-        xml = minidom.parseString(ET.tostring(xml_element))
+        create_dir(self.xml_dir)
+        xml = minidom.parseString(text)
```

Comment:
it looks like we still do not know why we are parsing the XML here.
I can already foresee a problem with this and I think we already hit it. I think that when an EBI submission failed in the past due to non-ascii (i.e. UTF-8) characters are because the above calls `tostring` have ascii as default (see [here](https://docs.python.org/2/library/xml.etree.elementtree.html?highlight=tostring#xml.etree.ElementTree.tostring)), so if we actually have an UTF-8 character, it will fail in the other functions and will never hit line 578, which is the one that "in theory" is writing the file correctly (i.e in UTF-8).
I think we should remove the parsing here and change the tostring calls in the other functions to include the correct encoding, or the other functions just return the minidom object and the only place where a `tostring` or a `toxml` call happens is in here. Whatever is easier works.